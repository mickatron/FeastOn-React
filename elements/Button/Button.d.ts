import * as React from 'react';



export interface ButtonProps {
  [key: string]: any;
  /** An element type to render as (string or function). */
  as?: any;
  /** Primary content. */
  children?: React.ReactNode;
  /** Additional classes. */
  className?: string | array | object;
  /** BEM modifier attributes.*/
  modifier?: string | array ;
  /** Width className */
  width?: string;
}

declare class Button extends React.Component<ButtonProps, {}> {
}

export default Button;