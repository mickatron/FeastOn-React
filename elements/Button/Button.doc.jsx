/* eslint-disable */
import React from 'react';
import Component from './Button';
import PropTypes from 'prop-types';
import asType from '../../proptypes/as';
import classNameType from '../../proptypes/className';

/**
 * Renders a standard `button`.
 * 
 * ** Sub Components **
 * * `ButtonIcon` as `Button.Icon`
 * 
 * @composedWith UiWrapper
*/
function Button(props) {
  return <Component {...props} />;
}
Button.propTypes = {
  /** The type of element to render the component as. */
  as: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
    PropTypes.symbol,
  ]),
  /** BEM modifier names to be applied to the elements wrapper.*/
  modifier: PropTypes.oneOfType([
    PropTypes.oneOf([
      'true',
      'warn',
      'false',
      'link',
      'outline',
      'pill',
      'full',
      'zeta',
      'epsilon',
      'gamma',
      'delta',
      'beta',
      'alpha',
    ]),
      PropTypes.arrayOf(PropTypes.oneOf( [
      'true',
      'warn',
      'false',
      'link',
      'outline',
      'pill',
      'full',
      'zeta',
      'epsilon',
      'gamma',
      'delta',
      'beta',
      'alpha',
    ])),
  ]),
};

Button.defaultProps = {
  as: 'button',  
  modifier: undefined,
  width: undefined,
};

export default Button;
