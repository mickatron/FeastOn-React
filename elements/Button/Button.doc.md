```js noeditor 
{
  const Button = require('./Button');
}
null
```
### Default Button style modifiers.
```js
<>
  <Button>Push Me</Button>
  <Button 
    className="test"
    modifier={["outline"]}
  >
    Push Me
  </Button>
  <Button 
    modifier={["pill"]}
  >
    Push Me
  </Button>
  <Button 
    modifier={["true"]}
  >
    Push Me
  </Button>
  <Button 
    modifier={["warn"]}
  >
    Push Me
  </Button>
  <Button 
    modifier={["false"]}
  >
    Push Me
  </Button>
  <Button 
    modifier={["link"]}
  >
    Push Me
  </Button>
</>
```
### Default Button size modifiers.
```js
<>
  <Button 
    modifier="zeta"
  >Push Me</Button>
  <Button 
    modifier="epsilon"
  >Push Me</Button>
  <Button 
    modifier="delta"
  >Push Me</Button>
  <Button 
    modifier="gamma"
  >Push Me</Button>
  <Button 
    modifier="beta"
  >Push Me</Button>
  <Button 
    modifier="alpha"
  >Push Me</Button>
</>
```
