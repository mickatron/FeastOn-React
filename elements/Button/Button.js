import uiWrapper from '../../HOC/uiWrapper/uiWrapper';

/**
 * return HOC wrapped component
 */

export default uiWrapper({
  name: 'Button',
  element: 'button',
  modifiers: [
    'true',
    'warn',
    'false',
    'link',
    'outline',
    'pill',
    'full',
    'zeta',
    'epsilon',
    'gamma',
    'delta',
    'beta',
    'alpha',
  ],
});
