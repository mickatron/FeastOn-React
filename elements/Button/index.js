import Button from './Button';
import ButtonIcon from './ButtonIcon';

Button.Icon = ButtonIcon;

export default Button;
