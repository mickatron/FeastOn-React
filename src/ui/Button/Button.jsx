import classnames from 'classnames';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import React from 'react';

import alphaSizesType from 'js/shared/proptypes/alphaSizes';
import classNameType from 'js/shared/proptypes/className';

export default class Button extends React.PureComponent { 
  
  static propTypes = {
    appearance: PropTypes.oneOf([
      'button',
      'link',
      'outline',
      'custom',
    ]),
    className: classNameType,
    children: PropTypes.node.isRequired,
    Icon:PropTypes.oneOfType([ 
      PropTypes.element,
      PropTypes.node,
    ]),
    iconPos: PropTypes.oneOf(['left','right']),
    mode: PropTypes.oneOf([
      'button',
      'link',
    ]),
    size: alphaSizesType,
  }

  static defaultProps = {
    appearance: 'button',
    Icon: undefined,
    iconPos: 'left',
    mode: 'button',
    size: 'delta',
  }

  renderChildren() {
    const {
      children,
      Icon,
      iconPos,
    } = this.props;
    return [
      Icon && iconPos==='left'? <Icon className="c-Button__icon" />: null ,
      children,
      Icon && iconPos==='right'? <Icon className="c-Button__icon" />: null,
    ];
  }

  render() { 
    const {
      appearance,
      className, 
      children,
      Icon,
      iconPos,
      mode,
      size,
      ...other
    } = this.props;
    const otherClassNames = {
      'c-Button--link': appearance === 'link' || false,
      'c-Button--outline': appearance === 'outline' || false,
    };

    const classes = classnames(className, 'c-Button', 'c-Button--' + size, otherClassNames);

    const renderedButton = {
      button: ()=>(
        <button 
          className={classes} 
          { ...other }
        >
          {this.renderChildren()}
        </button>
      ),
      link: ()=>(
        <Link 
          className={classes} 
          { ...other }
        >
          {this.renderChildren()}
        </Link>
      ),
    }[mode];

    return renderedButton();
  };
}
