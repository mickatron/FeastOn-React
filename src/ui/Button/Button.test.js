
import { MemoryRouter } from 'react-router-dom';
import { render, shallow } from 'enzyme';
import React from 'react';
import Button from './Button';

describe('Button', () => {

  let component = null;
  describe('renders', ()=>{
    it('custom className', () => {    
      component = shallow(
        <Button 
          className="test"
        >button text</Button>
      );
      expect(component).toMatchSnapshot();
      expect(component.find('.c-Button.test').exists()).toBeTruthy();
      
    });

    it('link button', () => {
      component = render(
        <MemoryRouter>
          <Button 
            mode="link"
            to="/"
          >button text</Button>
        </MemoryRouter>
      );
      expect(component).toMatchSnapshot();
    });

    it('outline button', () => {
      component = shallow(
        <Button 
          appearance="outline"
        >button text</Button>
      );
      expect(component).toMatchSnapshot();
      expect(component.find('.c-Button--outline').exists()).toBeTruthy();
    });

  });
});