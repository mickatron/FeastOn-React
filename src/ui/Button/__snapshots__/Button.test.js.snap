// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders custom className 1`] = `
<button
  className="test c-Button c-Button--delta"
>
  button text
</button>
`;

exports[`Button renders link button 1`] = `
<a
  class="c-Button c-Button--delta"
  href="/"
>
  button text
</a>
`;

exports[`Button renders outline button 1`] = `
<button
  className="c-Button c-Button--delta c-Button--outline"
>
  button text
</button>
`;
